### PR TITLE
:lipstick: Create Reusable Slider Button Component #24

### LINKED ISSUE #24: 🎨 Create Reusable Slider Button Component
#  Create a reusable Slider Button component
---

Quick links: [Summary](#summary) | [Description](#description) | [Technical-Analysis](#Technical-Analysis) | [Acceptance-Criteria](#Acceptance-Criteria)

---

## Summary

As a developer, I want a reusable slider button component that I can reuse anywhere it's needed across my application.

## Description

**Design**:

slider-button

slider-button-2

[Link to full design](https://xd.adobe.com/view/b98caf27-6549-49e5-8ed8-c5234788df61-ecdf/screen/cc35e19a-24f2-42ce-8803-ee2942839528)

### Package

The button component should be placed in the library **libs/elements/layout/buttons** (buttons is a module/library)

Use the command `nx generate @nrwl/angular:library elements/layout/buttons` to create this library. Make sure it's a separate commit.

The normal button component is a reusable component that can be reused across my application. The module is named "ButtonsModule". The normal button is being developed in #23. Avoid merge conflicts when creating the package.

The ButtonsModule exports the elewa-normal-button component, for other modules to import.

You are adding a second component to this package, i.e. elewa-slider-button.

### Input and outputs

The component exposes one Angular input of type:

```typeScript
interface SliderButtonData extends ButtonData
{
  text: string;

  bgColour?: string; // Background colour
  color?: string;
  borderColor?: string;
  iconColor?: string;
  iconBackgroundColor?: string;

  hoverBgColour?: string;
  hoverColor?: string;
  hoverBorderColor?: string;
  hoverIconColor?: string;
  hoverIconBackgroundColor?: string;
}
```
Most of the parameters are optional, and therefore nullable (hence the ?) on the variable. If the variables are set, the button adopts the correct colours. If not, the default colours are used as per the design spec.

The type/model should be stored in `libs/models/schema/ui/buttons` (buttons is a module/library) The name of the file containing the interface is slider-button-data.interface.ts and it's exported by its parent module above.

On click, the button exports an output called (click). The click output is an EventEmitter of type void.

### Behaviour

When hovering over the slider, the mouse shows the button is clickable and the slider adapts and moves via an animation. The icon at the end moves to the front and the colour changes per design spec. Use SCSS animations for this.

Default, the hover inverts the colours so if the icon is black and the slider is white, the slider becomes black and the icon becomes white.

### Acceptance Criteria

- [ ] When rendered the component is pixel-perfect compared to the design above.
- [ ] The component is clean and does not have more than 50 lines
- [ ] The component SCSS uses CSS variables for colour, font, ... Connected to the main style.scss
- [ ] The code is isolated to the `libs/elements/layout/buttons` package unless a justified reason is specified.
- [ ] The model/schema is isolated in the `libs/models/schema/ui/buttons` package unless a justified reason is specified.
- [ ] The design is responsive for screen, tablet and phone
- [ ] A separate commit is used for each small change
- [ ] The commits and PR are well formatted/use proper documentation
- [ ] The module exports the component

### PR DESCRIPTION
**Pull request by levy-web**
# Description

I have created a reusable slider button component that uses style directives to determine items style which is dependent on the one using the element .

Reference the issue related to this PR as shown below. Every issue has it's own number you can find the issue numbers [here](https://github.com/italanta/italanta-apps/issues).

Fixes #24

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Screenshots

**Reused the component to create two buttons on  Hover and not on Hover**
Screenshot from 2023-08-19 17-18-57

 **Video demonstration**

[Screencast from 19-08-2023  5:19:29 WB.webm](https://github.com/italanta/elewa-website/assets/71401172/147f67e2-0e34-4c1c-b01a-d58206072002)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**: (optional)

- Firmware version:
- Hardware:
- Toolchain:
- SDK:

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: apps/elewa-website/src/styles.scss [hunk 1]
```diff
@@ -1,42 +0,0 @@
-/* You can add global styles to this file, and also import other style files */
-
-@import '_variables';
-
-* {
-  box-sizing: border-box;
-  margin: 0;
-  padding: 0;
```
Comment (LemmyMwaura) [inline apps/elewa-website/src/styles.scss:8]: This removes the default styles from the site.

File: libs/elements/layout/buttons/src/lib/elewa-slider-button/elewa-slider-button.component.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+@Component({
+  selector: 'elewa-website-elewa-slider-button', // Component selector
+  templateUrl: './elewa-slider-button.component.html', // Template file for the component
+  styleUrls: ['./elewa-slider-button.component.scss'], // Styles file for the component
+})
+export class ElewaSliderButtonComponent {
+  @Input() sliderButtonData!: SliderButtonData; // Input data for the slider button
+  @Output() sliderButtonClick: EventEmitter<void> = new EventEmitter<void>(); // Output event for button click
+  hovered = false; // Flag to track hover state
```
Comment (LemmyMwaura) [inline libs/elements/layout/buttons/src/lib/elewa-slider-button/elewa-slider-button.component.ts:13]: For comments kindly use JsDoc comments, so we can get some good IntelliSense.

File: libs/features/pages/home/src/lib/main/home/home-page.component.ts [hunk 3]
```diff
@@ -1,8 +1,52 @@
...
+    iconColor: "black",
+    iconBackgroundColor: "white",
+  
+    hoverBgColour: "white",
+    hoverColor: "black",
+    hoverBorderColor: "",
+    hoverIconColor: "white",
+    hoverIconBackgroundColor: "black"
+
+  }
```
Comment (LemmyMwaura) [inline libs/features/pages/home/src/lib/main/home/home-page.component.ts:26]: No need to define any of this here.. Any component that needs this button will import it.

File: libs/models/schema/ui/buttons/src/lib/slider-button-data.interface.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+export interface ButtonData 
+{
+  text: string
+  
+  bgColor?: string; // Background colour
+  color?: string; // text color
+  
+  hoverBgColor?: string;
+  hoverColor?: string;
+}
```
Comment (LemmyMwaura) [inline libs/models/schema/ui/buttons/src/lib/slider-button-data.interface.ts:27]: Move this interface to a separate file and then import it.

File: tsconfig.base.json [hunk 5]
```diff
@@ -15,14 +15,17 @@
     "skipDefaultLibCheck": true,
     "baseUrl": ".",
     "paths": {
-      "@elewa-website/elements/layout/header": [
-        "libs/elements/layout/header/src/index.ts"
-      ],
```
Comment (LemmyMwaura) [inline tsconfig.base.json:20]: This removes the header aliased path from the codebase

**Review by LemmyMwaura — CHANGES_REQUESTED** [review]

File: tsconfig.base.json (hunk 5 above)
Comment (levy-web) [reply tsconfig.base.json:20]: @LemmyMwaura this path is still there it was not removed

File: libs/models/schema/ui/buttons/src/lib/slider-button-data.interface.ts (hunk 4 above)
Comment (levy-web) [reply libs/models/schema/ui/buttons/src/lib/slider-button-data.interface.ts]: the buttonData had not been created yet but I have used another interface file and imported the dummy interface

File: libs/features/pages/home/src/lib/main/home/home-page.component.ts (hunk 3 above)
Comment (levy-web) [reply libs/features/pages/home/src/lib/main/home/home-page.component.ts]: this was for my testing I wasnt supposed to push the code and I have removed the data

File: apps/elewa-website/src/styles.scss (hunk 1 above)
Comment (levy-web) [reply apps/elewa-website/src/styles.scss]: changed file to as of main repository

File: libs/elements/layout/buttons/src/lib/elewa-slider-button/elewa-slider-button.component.ts (hunk 2 above)
Comment (levy-web) [reply libs/elements/layout/buttons/src/lib/elewa-slider-button/elewa-slider-button.component.ts]: updated and pushed the code with jsdoc comments

**Comment by levy-web** [comment]
@LemmyMwaura I'm done with requested changes. kindly review

File: tsconfig.base.json (hunk 5 above)
Comment (LemmyMwaura) [reply tsconfig.base.json:20]: Okay, I've seen it.

**Comment by LemmyMwaura** [comment]
Good work. works as expected but I've gone with PR #67 which was completed first as per my initial review. Really looking forward to your next contribution.

**Pull request closed without merging**